pennylane/tape/qscript.py

```diff
@@ -0,0 +1,1282 @@
+    "U2": "u2",
+    "U1": "u1",
+    "Identity": "id",
+    "PauliX": "x",
+    "PauliY": "y",
+    "PauliZ": "z",
+    "Hadamard": "h",
+    "S": "s",
+    "S.inv": "sdg",
+    "T": "t",
+    "T.inv": "tdg",
+    "RX": "rx",
+    "RY": "ry",
+    "RZ": "rz",
+    "CRX": "crx",
+    "CRY": "cry",
+    "CRZ": "crz",
+    "SWAP": "swap",
+    "Toffoli": "ccx",
+    "CSWAP": "cswap",
+    "PhaseShift": "u1",
+}
+"""
+dict[str, str]: Maps PennyLane gate names to equivalent QASM gate names.
+
+Note that QASM has two native gates:
+
+- ``U`` (equivalent to :class:`~.U3`)
+- ``CX`` (equivalent to :class:`~.CNOT`)
+
+All other gates are defined in the file stdgates.inc:
+https://github.com/Qiskit/openqasm/blob/master/examples/stdgates.inc
+"""
+
+
+class QuantumScript:
+    """The state preparation, operations, and measurments that represent instructions for
+    execution on a quantum device.
+
+    Args:
+        ops (Iterable[Operator]): An iterable of the operations to be performed
+        measurements (Iterable[MeasurementProcess]): All the measurements to be performed
+        prep (Iterable[Operator]): Any state preparations to perform at the start of the circuit
+
+    Keyword Args:
+        name (str): a name given to the quantum script
+        do_queue=False (bool): Whether or not to queue. Defaults to ``True`` for ``QuantumTape``.
+        _update=True (bool): Whether or not to set various properties on initialization. Setting
+            ``_update=False`` reduces computations if the script is only an intermediary step.
+
+    .. seealso:: :class:`pennylane.tape.QuantumTape`
+
+    **Example:**
+
+    .. code-block:: python
+
+        from pennylane.tape import QuantumScript
+
+        prep = [qml.BasisState([1,1], wires=(0,1))]
+
+        ops = [qml.RX(0.432, 0),
+               qml.RY(0.543, 0),
+               qml.CNOT((0,"a")),
+               qml.RX(0.133, "a")]
+
+        qscript = QuantumScript(ops, [qml.expval(qml.PauliZ(0))], prep)
+
+    >>> list(qscript)
+    [RX(0.432, wires=[0]),
+    RY(0.543, wires=[0]),
+    CNOT(wires=[0, 'a']),
+    RX(0.133, wires=['a']),
+    expval(PauliZ(wires=[0]))]
+    >>> qscript.operations
+    [RX(0.432, wires=[0]),
+    RY(0.543, wires=[0]),
+    CNOT(wires=[0, 'a']),
+    RX(0.133, wires=['a'])]
+    >>> qscript.measurements
+    [expval(PauliZ(wires=[0]))]
+
+    Iterating over the quantum script can be done by:
+
+    >>> for op in qscript:
+    ...     print(op)
+    RX(0.432, wires=[0])
+    RY(0.543, wires=[0])
+    CNOT(wires=[0, 'a'])
+    RX(0.133, wires=['a'])
+    expval(PauliZ(wires=[0]))'
+
+    Quantum scripts also support indexing and length determination:
+
+    >>> qscript[0]
+    RX(0.432, wires=[0])
+    >>> len(qscript)
+    5
+
+    Once constructed, the script can be executed directly on a quantum device
+    using the :func:`~.pennylane.execute` function:
+
+    >>> dev = qml.device('default.qubit', wires=(0,'a'))
+    >>> qml.execute([qscript], dev, gradient_fn=None)
```

Comment:
Running this raises an error....

```
File ~/Developer/pennylane/pennylane/devices/default_qubit.py:731, in DefaultQubit._apply_basis_state(self, state, wires)
    728 # length of basis state parameter
    729 n_basis_state = len(state)
--> 731 if not set(state.tolist()).issubset({0, 1}):
    732     raise ValueError("BasisState parameter must consist of 0 or 1 integers.")
    734 if n_basis_state != len(device_wires):

AttributeError: 'list' object has no attribute 'tolist'
```

Reply:
This was due to the BasisState having a list instead of a numpy array. Should be fixed now.